spatial/curve/hilbert.go

```diff
@@ -0,0 +1,241 @@
+		rx := (e>>0 ^ e>>1) & 1
+		v[0] += rx << n
+		v[1] += ry << n
+		v[2] += rz << n
+		pos >>= 3
+	}
+}
+
+// Coord returns the spatial coordinates of pos as a slice.
+func (h Hilbert3D) Coord(pos int) []int {
+	v := make([]int, 3)
+	h.WriteCoordTo(pos, v)
+	return v[:]
+}
+
+// Hilbert4D is a 4-dimensional Hilbert curve.
+type Hilbert4D struct{ Order int }
+
+// Dims returns the spatial dimensions of the curve, which is {2ᵏ, 2ᵏ, 2ᵏ, 2ᵏ},
+// where k is the order.
+func (h Hilbert4D) Dims() []int { return []int{1 << h.Order, 1 << h.Order, 1 << h.Order, 1 << h.Order} }
+
+// Len returns the length of the curve, which is 2ⁿᵏ, where n is the dimension
+// (4) and k is the order.
+//
+// Len will overflow on a 32-bit architecture if the order is ≥ 8.
+func (h Hilbert4D) Len() int { return 1 << (4 * h.Order) }
+
+func (h Hilbert4D) rot(reverse bool, n int, v []int, d int) {
+	switch d {
+	case 0:
+		do2(reverse, n, v, swap{1, 3}, swap{0, 3})
+	case 1, 2:
+		do2(reverse, n, v, swap{0, 3}, swap{1, 3})
+	case 3, 4:
+		do2(reverse, n, v, flip{0, 1}, swap{2, 3})
+	case 5, 6:
+		do2(reverse, n, v, flip{1, 2}, swap{2, 3})
+	case 7, 8:
+		invert{0, 2}.do(n, v)
+	case 9, 10:
+		do2(reverse, n, v, flip{1, 2}, flip{2, 3})
+	case 11, 12:
+		do2(reverse, n, v, flip{0, 1}, flip{2, 3})
+	case 13, 14:
+		do2(reverse, n, v, flip{0, 3}, flip{1, 3})
+	case 15:
+		do2(reverse, n, v, flip{1, 3}, flip{0, 3})
+	}
+}
+
+// Pos returns the linear position of the spatial coordinate along the curve.
+// Pos modifies v.
+//
+// Pos will overflow on a 32-bit architecture if the order is ≥ 8.
+func (h Hilbert4D) Pos(v []int) int {
+	var d int
+	N := 4
+	for n := h.Order - 1; n >= 0; n-- {
+		var e int
+		for i := N - 1; i >= 0; i-- {
+			v := v[i] >> n & 1
+			e = e<<1 | (e^v)&1
+		}
+
+		d += e << (N * n)
+		h.rot(false, h.Order, v, e)
+	}
+	return d
+}
+
+// WriteCoordTo writes the spatial coordinates of pos to v.
+func (h Hilbert4D) WriteCoordTo(pos int, v []int) {
+	N := 4
+	for n := 0; n < h.Order; n++ {
+		e := pos & (1<<N - 1)
+		h.rot(true, n, v[:], e)
+
+		for i, e := 0, e; i < N; i++ {
+			v[i] += (e ^ e>>1) & 1 << n
+			e >>= 1
+		}
+		pos >>= N
+	}
+}
+
+// Coord returns the spatial coordinates of pos as a slice.
+func (h Hilbert4D) Coord(pos int) []int {
+	v := make([]int, 4)
+	h.WriteCoordTo(pos, v)
+	return v
+}
+
+type op interface{ do(int, []int) }
+
+// invert I and J
+type invert struct{ i, j int }
+
+func (c invert) do(n int, v []int) { v[c.i], v[c.j] = v[c.i]^(1<<n-1), v[c.j]^(1<<n-1) }
+
+// swap I and J
+type swap struct{ i, j int }
+
+func (c swap) do(n int, v []int) { v[c.i], v[c.j] = v[c.j], v[c.i] }
+
+// swap and invert I and J
+type flip struct{ i, j int }
+
+func (c flip) do(n int, v []int) { v[c.i], v[c.j] = v[c.j]^(1<<n-1), v[c.i]^(1<<n-1) }
+
+// do2 executes the given operations, optionally in reverse.
+//
+// Generic specialization reduces allocation (because it can eliminate interface
+// value boxing) and improves performance
```

Comment:
Nice that you found this. I was not aware that this was the case.

Reply:
I knew that `var x any = valueType` results in an allocation (if `x` escapes to the heap) and I could see in profiles that `ops.do` was a significant source of allocations so I inferred that boxing the operations structs as `type op inferface` was the source of those allocations. Though (the reference implementation of) Go uses gc-shape instead of true templating (as in C++), the resulting machine code is identical to a templated version I believe, so I figured this would work. Having two separate type parameters for the two function parameters means that the end result is identical to defining a separate `do2` function hard coded for each combination we use. I *think* there's only one actual do2 implementation generated since all the operations have the same shape, but it's possible the compiler is generating a separate implementation for each combination.